### PR TITLE
set logger in constructor

### LINKED ISSUE #17925: Encryption migration 500 8.0.4->stable8.1 (upcoming 8.1.1)
### Steps

1. Setup OC 8.0.4
1. Enable encryption app
1. Log out then log in again
1. Create two users "user1" and "user2"
1. Login as "user1"
1. Create a folder "test"
1. Share "test" with "user2"
1. Share "test" with link
1. Upgrade the server to stable8.1 (a949f6c5762c12f53559fb4329a9ec9e7ce5b415), the upcoming 8.1.1
1. Login as admin
1. Enable "default encryption module"
1. Go to admin page, "Enable server side encryption", "Enable encryption"
1. Click "Start migration"

### Expected result
Migration runs.

### Actual result
500 error:
```
{"reqId":"BqFryMWB9VISbs2D3hdM","remoteAddr":"127.0.0.1","app":"no app in context","message":"Installation is in transit between the old Encryption (ownCloud <= 8.0)\n\t\t\tand the new encryption. Please enable the \"Default encryption module\"\n\t\t\tand run 'occ encryption:migrate'","level":2,"time":"2015-07-28T11:06:16+00:00","method":"POST","url":"\/owncloud\/index.php\/settings\/admin\/startmigration"}
{"reqId":"BqFryMWB9VISbs2D3hdM","remoteAddr":"127.0.0.1","app":"PHP","message":"Argument 4 passed to OCA\\Encryption\\Migration::__construct() must implement interface OCP\\ILogger, null given, called in \/srv\/www\/htdocs\/owncloud\/settings\/controller\/encryptioncontroller.php on line 94 and defined at \/srv\/www\/htdocs\/owncloud\/apps\/encryption\/lib\/migration.php#50","level":3,"time":"2015-07-28T11:06:16+00:00","method":"POST","url":"\/owncloud\/index.php\/settings\/admin\/startmigration"}
{"reqId":"BqFryMWB9VISbs2D3hdM","remoteAddr":"127.0.0.1","app":"PHP","message":"Call to a member function info() on null at \/srv\/www\/htdocs\/owncloud\/apps\/encryption\/lib\/migration.php#273","level":3,"time":"2015-07-28T11:06:16+00:00","method":"POST","url":"\/owncloud\/index.php\/settings\/admin\/startmigration"}
```

@schiesbn looks like some API mismatch ?

### PR DESCRIPTION
**Pull request by schiessle**
fix #17925 

@PVince81 please give it a try, this should fix the issue you reported.
@karlitschek this needs to be backported to 8.1.1

### CODE REVIEW

**Comment by LukasReschke** [comment]
Shouldn't this be unit testable?

**Comment by karlitschek** [comment]
please backport :+1:

**Comment by Xenopathic** [comment]
:+1:

@LukasReschke probably, but is it really worth unit testing every constructor call that is simply setting class members? Integration tests will pick up on issues like this anyway

**Comment by LukasReschke** [comment]
> @LukasReschke probably, but is it really worth unit testing every constructor call that is simply setting class members? Integration tests will pick up on issues like this anyway

I am not talking about unit testing every constructor. I am talking about actually unit testing the class. From my understanding this would have failed since `$this->logger` is not defined leading to a:

> Argument 4 passed to OCA\Encryption\Migration::__construct() must be an instance of \OCP\ILogger, null given

**Comment by Xenopathic** [comment]
@LukasReschke Ahh, forgot about that. Yeah, where's the unit tests for this class? :speak_no_evil:

**Comment by LukasReschke** [comment]
… I will write unit tests…

**Comment by PVince81** [comment]
:+1: tested works.

Curious to see how the unit test will look like.

**Comment by LukasReschke** [comment]
Unit tests added. They catch this error at 4 possible places.

**Comment by PVince81** [comment]
Thanks @LukasReschke. I will retest.

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please

**Comment by PVince81** [comment]
Still works :+1:

**Comment by Xenopathic** [comment]
Tests pass, good code :+1:

File: settings/controller/encryptioncontroller.php [hunk 1]
```diff
@@ -78,6 +78,24 @@ public function __construct($appName,
...
+	 * Public for unit testing purposes
+	 *
+	 * @internal
+	 * @param IConfig $config
+	 * @param View $view
+	 * @param Connection $connection
+	 * @param ILogger $logger
+	 * @return Migration
+	 */
+	public function getMigration(IConfig $config,
```
Comment (schiesbn) [inline settings/controller/encryptioncontroller.php]: you made it public just to mock the method right? No need for it... protected works as well. Please don't expose methods just for unit testing.

File: settings/controller/encryptioncontroller.php [hunk 2]
```diff
@@ -78,6 +78,24 @@ public function __construct($appName,
 		$this->connection = $connection;
 		$this->view = $view;
 		$this->userManager = $userManager;
+		$this->logger = $logger;
+	}
+
+	/**
+	 * Public for unit testing purposes
```
Comment (schiesbn) [inline settings/controller/encryptioncontroller.php]: comment no longer match :speak_no_evil:

**Comment by schiesbn** [comment]
@lukas thanks for the unit test. I also added it to the backport.

**Comment by schiesbn** [comment]
@owncloud-bot retest this please

File: settings/controller/encryptioncontroller.php (hunk 1 above)
Comment (LukasReschke) [inline settings/controller/encryptioncontroller.php]: Indeed. That makes much more sense :speak_no_evil:

**Pull request merged**